<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>CaSR-R568 RNA-seq</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="CaSR_files/libs/clipboard/clipboard.min.js"></script>
<script src="CaSR_files/libs/quarto-html/quarto.js"></script>
<script src="CaSR_files/libs/quarto-html/popper.min.js"></script>
<script src="CaSR_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="CaSR_files/libs/quarto-html/anchor.min.js"></script>
<link href="CaSR_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="CaSR_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="CaSR_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="CaSR_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="CaSR_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">CaSR-R568 RNA-seq</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p><em>Paper:</em> <a href="https://doi.org/10.1016/j.bbamcr.2020.118836">Iamartino et al.&nbsp;Effects of pharmacological calcimimetics on colorectal cancer cells over-expressing the human calcium-sensing receptor. 2020</a>.</p>
<p><em>Studying the transcriptome effect of Calcium sensing receptor positive modulator R568 by RNA-seq.</em></p>
<p><em>Tutorials followed:</em> <a href="https://f1000research.com/articles/5-1408">Law et al.&nbsp;RNA-seq analysis is easy as 1-2-3 with limma, Glimma and edgeR</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">browseVignettes</span>(<span class="st">"edgeR"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">browseVignettes</span>(<span class="st">"limma"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="data-import" class="level1">
<h1>Data import</h1>
<p>Download raw count from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE140984">GSE140984</a>.</p>
<p>Import</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="fu">list.files</span>(<span class="at">pattern =</span> <span class="st">".tsv"</span>))</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(counts) <span class="ot">&lt;-</span> counts[,<span class="dv">1</span>] <span class="co"># first column is Ensembl gene ID</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> counts[,<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  EK02001 EK02002 EK02003 EK02004 EK02005 EK02006 EK02007
ENSG00000223972.5       2       0       0       0       0       1       0
ENSG00000227232.5       0       1       2       2       0       2       1
ENSG00000278267.1       2       4       5       5       3       1       2
ENSG00000243485.5       0       0       0       0       0       0       0
ENSG00000284332.1       0       0       0       0       0       0       0
ENSG00000237613.2       0       0       0       0       0       0       0
                  EK02008 EK02009 EK02010 EK02011 EK02012
ENSG00000223972.5       1       0       1       0       0
ENSG00000227232.5       0       0       0       0       0
ENSG00000278267.1       2       6       0       4       4
ENSG00000243485.5       0       0       0       0       0
ENSG00000284332.1       0       0       0       0       0
ENSG00000237613.2       0       0       0       0       0</code></pre>
</div>
</div>
<p>Make sample information</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">factor</span>(<span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"empty"</span>, <span class="dv">6</span>), <span class="fu">rep</span>(<span class="st">"CaSR"</span>, <span class="dv">6</span>))),</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">factor</span>(<span class="fu">c</span>(<span class="fu">rep</span>(<span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"DMSO"</span>, <span class="dv">3</span>), <span class="fu">rep</span>(<span class="st">"R568"</span>, <span class="dv">3</span>)), <span class="dv">2</span>)))</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(samples) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(counts)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(samples) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GFP"</span>, <span class="st">"treatment"</span>)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>samples<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">paste0</span>(samples<span class="sc">$</span>GFP, <span class="st">"."</span>, samples<span class="sc">$</span>treatment))</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>samples<span class="sc">$</span>GFP <span class="ot">&lt;-</span> <span class="fu">relevel</span>(samples<span class="sc">$</span>GFP, <span class="st">"empty"</span>)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          GFP treatment      group
EK02001 empty      DMSO empty.DMSO
EK02002 empty      DMSO empty.DMSO
EK02003 empty      DMSO empty.DMSO
EK02004 empty      R568 empty.R568
EK02005 empty      R568 empty.R568
EK02006 empty      R568 empty.R568
EK02007  CaSR      DMSO  CaSR.DMSO
EK02008  CaSR      DMSO  CaSR.DMSO
EK02009  CaSR      DMSO  CaSR.DMSO
EK02010  CaSR      R568  CaSR.R568
EK02011  CaSR      R568  CaSR.R568
EK02012  CaSR      R568  CaSR.R568</code></pre>
</div>
</div>
<p>Make gene annotation</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> <span class="fu">substr</span>(<span class="fu">rownames</span>(counts), <span class="dv">1</span>, <span class="dv">15</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>symbol <span class="ot">&lt;-</span> <span class="fu">mapIds</span>(org.Hs.eg.db,</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">keys =</span> ensembl, </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">column =</span> <span class="st">"SYMBOL"</span>,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>entrez <span class="ot">&lt;-</span> <span class="fu">mapIds</span>(org.Hs.eg.db,</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>                 <span class="at">keys =</span> ensembl, </span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>                 <span class="at">column =</span> <span class="st">"ENTREZID"</span>,</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>                 <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">symbol =</span> symbol,</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>                    <span class="at">entrezid =</span> entrez)</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(genes) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(counts)</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                     symbol  entrezid
ENSG00000223972.5      &lt;NA&gt;      &lt;NA&gt;
ENSG00000227232.5      &lt;NA&gt;      &lt;NA&gt;
ENSG00000278267.1 MIR6859-1 102466751
ENSG00000243485.5      &lt;NA&gt;      &lt;NA&gt;
ENSG00000284332.1 MIR1302-2 100302278
ENSG00000237613.2   FAM138A    645520</code></pre>
</div>
</div>
</section>
<section id="make-dgelist-object-and-preprocess" class="level1">
<h1>Make DGEList object and preprocess</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(edgeR)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(<span class="at">counts =</span> counts,</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">samples =</span> samples,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">genes =</span> genes)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 58785    12</code></pre>
</div>
</div>
<p>Filter low count genes and genes without annotations</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge, <span class="at">group =</span> samples<span class="sc">$</span>group)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(keep)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Mode   FALSE    TRUE 
logical   43690   15095 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>hasannot <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">is.na</span>(dge<span class="sc">$</span>genes)) <span class="sc">==</span> <span class="dv">0</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(hasannot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Mode   FALSE    TRUE 
logical   22736   36049 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> dge[keep <span class="sc">&amp;</span> hasannot, , keep.lib.sizes <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 13400    12</code></pre>
</div>
</div>
<p>TMM normalization</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>log2cpm <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dge, <span class="at">log =</span> <span class="cn">TRUE</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(log2cpm, <span class="at">las =</span> <span class="dv">2</span>) <span class="co"># before normalization, not much variation</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">normLibSizes</span>(dge) <span class="co"># newer function name for calcNormFactors</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>log2cpm <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dge, <span class="at">log =</span> <span class="cn">TRUE</span>)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(log2cpm, <span class="at">las =</span> <span class="dv">2</span>) <span class="co"># after normalization</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Examine MD plots of each samples</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">4</span>))</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>){</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plotMD</span>(log2cpm, <span class="at">column =</span> i)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">0</span>, <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>} <span class="co"># The counts are not too spread out from center indicating good normalization</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="data-exploration" class="level1">
<h1>Data exploration</h1>
<p>Unsupervised clustering by MDS plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>points <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">15</span>,<span class="dv">16</span>,<span class="dv">0</span>,<span class="dv">1</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"purple"</span>, <span class="st">"red"</span>, <span class="st">"gray"</span>, <span class="st">"green"</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(log2cpm,</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">labels =</span> samples<span class="sc">$</span>group,</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> colors[samples<span class="sc">$</span>group]) <span class="co"># We can see separations of CaSR+ samples</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<p>Make design matrix according to group</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> samples<span class="sc">$</span>group)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(design) <span class="ot">&lt;-</span> <span class="fu">levels</span>(samples<span class="sc">$</span>group)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>design</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   CaSR.DMSO CaSR.R568 empty.DMSO empty.R568
1          0         0          1          0
2          0         0          1          0
3          0         0          1          0
4          0         0          0          1
5          0         0          0          1
6          0         0          0          1
7          1         0          0          0
8          1         0          0          0
9          1         0          0          0
10         0         1          0          0
11         0         1          0          0
12         0         1          0          0
attr(,"assign")
[1] 1 1 1 1
attr(,"contrasts")
attr(,"contrasts")$`samples$group`
[1] "contr.treatment"</code></pre>
</div>
</div>
<p>Estimate dispersion</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">estimateDisp</span>(dge, design, <span class="at">robust =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotBCV</span>(dge) <span class="co"># Dispersion trend increases with counts then smooth out as expected</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Squeeze dispersion to trend by QL (quasi-likelihood) approach to improve power</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">glmQLFit</span>(dge, design, <span class="at">robust =</span> <span class="cn">TRUE</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotQLDisp</span>(fit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="differential-expression" class="level1">
<h1>Differential expression</h1>
<section id="edger-pipeline" class="level2">
<h2 class="anchored" data-anchor-id="edger-pipeline">edgeR pipeline</h2>
<p>We test CaSR+ samples only</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>con <span class="ot">&lt;-</span> <span class="fu">makeContrasts</span>(<span class="at">treated =</span> CaSR.R568 <span class="sc">-</span> empty.DMSO,</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">untreated =</span> CaSR.DMSO <span class="sc">-</span> empty.DMSO,</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">levels =</span> design)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>qlf <span class="ot">&lt;-</span> <span class="fu">glmQLFTest</span>(fit, <span class="at">contrast =</span> con)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="fu">topTags</span>(qlf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Coefficient:  LR test on 2 degrees of freedom 
                    symbol entrezid logFC.treated logFC.untreated     logCPM
ENSG00000196352.15    CD55     1604      3.082078      0.40335582  6.1508544
ENSG00000109610.5     SOD3     6649      4.947141     -0.58564694  4.1988254
ENSG00000163762.6  TM4SF18   116441      5.529888      0.67437696 -0.5892991
ENSG00000143839.14     REN     5972      5.557782      0.64833926  5.3044671
ENSG00000265107.2     GJA5     2702      3.040164      0.33678311  2.6734750
ENSG00000173546.7    CSPG4     1464      2.665717      0.07895494  4.7507890
ENSG00000116741.7     RGS2     5997      2.497201     -0.32472159  4.9542767
ENSG00000213626.12     LBH    81606      5.473781      0.70922934  2.1409293
ENSG00000184371.13    CSF1     1435      3.588479      0.25092146  2.0041255
ENSG00000173175.14   ADCY5      111      4.162491     -0.23295025  2.2818813
                           F       PValue          FDR
ENSG00000196352.15 359.34689 1.302890e-11 1.028561e-07
ENSG00000109610.5  352.62774 1.535166e-11 1.028561e-07
ENSG00000163762.6   84.82781 1.129017e-10 5.042943e-07
ENSG00000143839.14 218.81740 2.654305e-10 8.891921e-07
ENSG00000265107.2  158.06075 1.789104e-09 4.794798e-06
ENSG00000173546.7  150.83924 2.394525e-09 5.347772e-06
ENSG00000116741.7  144.46444 3.085810e-09 5.907122e-06
ENSG00000213626.12 139.75595 4.434347e-09 7.427532e-06
ENSG00000184371.13 133.26809 5.014537e-09 7.466088e-06
ENSG00000173175.14 127.85544 6.453825e-09 8.648126e-06</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">decideTests</span>(qlf))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       LR test on 2 degrees of freedom
NotSig                           12871
Sig                                529</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMD</span>(qlf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>tr <span class="ot">&lt;-</span> <span class="fu">glmTreat</span>(fit, <span class="at">contrast =</span> con[,<span class="dv">1</span>], <span class="at">lfc =</span> <span class="dv">1</span>) <span class="co"># Test for genes with log2FC &gt; 1 for CaSR.R568</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">topTags</span>(tr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Coefficient:  1*CaSR.R568 -1*empty.DMSO 
                    symbol entrezid    logFC unshrunk.logFC     logCPM
ENSG00000196352.15    CD55     1604 3.082078       3.082362  6.1508544
ENSG00000109610.5     SOD3     6649 4.947141       4.950621  4.1988254
ENSG00000143839.14     REN     5972 5.557782       5.560256  5.3044671
ENSG00000163762.6  TM4SF18   116441 5.529888       5.699063 -0.5892991
ENSG00000213626.12     LBH    81606 5.473781       5.495137  2.1409293
ENSG00000265107.2     GJA5     2702 3.040164       3.043320  2.6734750
ENSG00000184371.13    CSF1     1435 3.588479       3.595571  2.0041255
ENSG00000173546.7    CSPG4     1464 2.665717       2.666300  4.7507890
ENSG00000173175.14   ADCY5      111 4.162491       4.170591  2.2818813
ENSG00000204866.8    IGFL2   147920 4.593635       4.614583  1.3222923
                         PValue          FDR
ENSG00000196352.15 2.950016e-10 2.591162e-06
ENSG00000109610.5  3.867407e-10 2.591162e-06
ENSG00000143839.14 2.262573e-09 1.010616e-05
ENSG00000163762.6  1.242292e-08 4.161677e-05
ENSG00000213626.12 4.166943e-08 1.050410e-04
ENSG00000265107.2  4.703328e-08 1.050410e-04
ENSG00000184371.13 9.743074e-08 1.865103e-04
ENSG00000173546.7  1.403375e-07 2.107223e-04
ENSG00000173175.14 1.415299e-07 2.107223e-04
ENSG00000204866.8  1.932354e-07 2.585800e-04</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">decideTests</span>(tr))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       1*CaSR.R568 -1*empty.DMSO
Down                          10
NotSig                     13321
Up                            69</code></pre>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMD</span>(tr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-13-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="limma-voom-pipeline" class="level2">
<h2 class="anchored" data-anchor-id="limma-voom-pipeline">limma-voom pipeline</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>v <span class="ot">&lt;-</span> <span class="fu">voom</span>(dge, design, <span class="at">plot =</span> <span class="cn">TRUE</span>) <span class="co"># genes with high count but not decrease with variance means biological variation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>vfit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(v, design)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>vfit <span class="ot">&lt;-</span> <span class="fu">contrasts.fit</span>(vfit, <span class="at">contrasts =</span> con)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>efit <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(vfit)</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSA</span>(efit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-14-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">topTable</span>(efit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   symbol entrezid  treated   untreated   AveExpr         F
ENSG00000196352.15   CD55     1604 3.067352  0.40189054 5.4313474 308.24478
ENSG00000109610.5    SOD3     6649 4.985474 -0.54208694 2.2389944 271.05054
ENSG00000143839.14    REN     5972 5.583567  0.62444548 3.1871119 187.91855
ENSG00000173546.7   CSPG4     1464 2.643247  0.06965275 4.2099504 126.09218
ENSG00000116741.7    RGS2     5997 2.485276 -0.32004639 4.3952274 124.49929
ENSG00000265107.2    GJA5     2702 3.020251  0.33862286 1.9987237 120.66694
ENSG00000184371.13   CSF1     1435 3.586095  0.23943873 1.0721989  93.52075
ENSG00000148671.13  ADIRF    10974 2.624742  0.72897676 2.7893577  90.29209
ENSG00000103018.16  CYB5B    80777 1.042612  0.06157945 7.5191283  81.40370
ENSG00000173175.14  ADCY5      111 4.395487 -0.07140604 0.8648382  81.01601
                        P.Value    adj.P.Val
ENSG00000196352.15 3.548978e-11 4.755631e-07
ENSG00000109610.5  7.655526e-11 5.129203e-07
ENSG00000143839.14 6.751924e-10 3.015859e-06
ENSG00000173546.7  7.025748e-09 2.026510e-05
ENSG00000116741.7  7.565422e-09 2.026510e-05
ENSG00000265107.2  9.073923e-09 2.026510e-05
ENSG00000184371.13 3.950730e-08 7.562827e-05
ENSG00000148671.13 4.830851e-08 8.091675e-05
ENSG00000103018.16 8.719770e-08 1.125679e-04
ENSG00000173175.14 8.959362e-08 1.125679e-04</code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">decideTests</span>(efit))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       treated untreated
Down       193         7
NotSig   12742     13393
Up         465         0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>tfit <span class="ot">&lt;-</span> <span class="fu">treat</span>(vfit, <span class="at">lfc =</span> <span class="dv">1</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">topTreat</span>(tfit, <span class="at">coef =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   symbol entrezid    logFC     AveExpr         t      P.Value
ENSG00000109610.5    SOD3     6649 4.985474 2.238994393 15.455842 1.109233e-09
ENSG00000196352.15   CD55     1604 3.067352 5.431347421 15.352673 1.189962e-09
ENSG00000143839.14    REN     5972 5.583567 3.187111872 14.412421 2.513732e-09
ENSG00000265107.2    GJA5     2702 3.020251 1.998723689  9.415117 2.983058e-07
ENSG00000213626.12    LBH    81606 5.652324 0.001291519  8.798180 6.296540e-07
ENSG00000184371.13   CSF1     1435 3.586095 1.072198929  8.770247 6.401516e-07
ENSG00000173546.7   CSPG4     1464 2.643247 4.209950420  8.649984 7.402983e-07
ENSG00000173175.14  ADCY5      111 4.395487 0.864838207  8.462230 9.392721e-07
ENSG00000148671.13  ADIRF    10974 2.624742 2.789357717  7.992650 1.697854e-06
ENSG00000138131.3   LOXL4    84171 2.838402 6.369975893  7.979539 1.727292e-06
                      adj.P.Val
ENSG00000109610.5  7.972748e-06
ENSG00000196352.15 7.972748e-06
ENSG00000143839.14 1.122800e-05
ENSG00000265107.2  9.993246e-04
ENSG00000213626.12 1.417142e-03
ENSG00000184371.13 1.417142e-03
ENSG00000173546.7  1.417142e-03
ENSG00000173175.14 1.573281e-03
ENSG00000148671.13 2.314571e-03
ENSG00000138131.3  2.314571e-03</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">decideTests</span>(tfit))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       treated untreated
Down         4         1
NotSig   13361     13399
Up          35         0</code></pre>
</div>
</div>
</section>
<section id="voom-with-quality-weights-pipeline" class="level2">
<h2 class="anchored" data-anchor-id="voom-with-quality-weights-pipeline">voom with quality weights pipeline</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>vwts <span class="ot">&lt;-</span> <span class="fu">voomWithQualityWeights</span>(dge, design, <span class="at">plot =</span> <span class="cn">TRUE</span>) <span class="co"># big difference in quality weights between samples</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>vwfit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(vwts)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>vwfit <span class="ot">&lt;-</span> <span class="fu">contrasts.fit</span>(vwfit, <span class="at">contrasts =</span> con)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>vwfit <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(vwfit)</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="fu">topTable</span>(vwfit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   symbol entrezid  treated   untreated  AveExpr        F
ENSG00000109610.5    SOD3     6649 4.936777 -0.56164614 2.238994 357.4661
ENSG00000196352.15   CD55     1604 2.967435  0.41305396 5.431347 347.8701
ENSG00000143839.14    REN     5972 5.476634  0.70025172 3.187112 285.0994
ENSG00000265107.2    GJA5     2702 3.039117  0.36559283 1.998724 195.6607
ENSG00000173546.7   CSPG4     1464 2.606368  0.13683996 4.209950 168.3871
ENSG00000116741.7    RGS2     5997 2.364882 -0.27652462 4.395227 134.2200
ENSG00000184371.13   CSF1     1435 3.485613  0.30409081 1.072199 115.5017
ENSG00000103018.16  CYB5B    80777 1.007659  0.08433645 7.519128 108.9104
ENSG00000148671.13  ADIRF    10974 2.534308  0.77219977 2.789358 108.5031
ENSG00000157193.16   LRP8     7804 1.471241  0.08169918 8.830237 105.5538
                        P.Value    adj.P.Val
ENSG00000109610.5  1.551301e-11 1.223111e-07
ENSG00000196352.15 1.825539e-11 1.223111e-07
ENSG00000143839.14 5.988362e-11 2.674802e-07
ENSG00000265107.2  5.585909e-10 1.871279e-06
ENSG00000173546.7  1.351906e-09 3.623107e-06
ENSG00000116741.7  5.092626e-09 1.137353e-05
ENSG00000184371.13 1.217226e-08 2.330119e-05
ENSG00000103018.16 1.708724e-08 2.599628e-05
ENSG00000148671.13 1.746019e-08 2.599628e-05
ENSG00000157193.16 2.046268e-08 2.742000e-05</code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">decideTests</span>(vwfit))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       treated untreated
Down       374         8
NotSig   12275     13391
Up         751         1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>twfit <span class="ot">&lt;-</span> <span class="fu">treat</span>(vwfit, <span class="at">lfc =</span> <span class="dv">1</span>)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">topTreat</span>(twfit, <span class="at">coef =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                    symbol entrezid    logFC      AveExpr         t
ENSG00000109610.5     SOD3     6649 4.936777  2.238994393 17.746844
ENSG00000143839.14     REN     5972 5.476634  3.187111872 17.732886
ENSG00000196352.15    CD55     1604 2.967435  5.431347421 16.069519
ENSG00000265107.2     GJA5     2702 3.039117  1.998723689 12.031522
ENSG00000173546.7    CSPG4     1464 2.606368  4.209950420 10.012707
ENSG00000163762.6  TM4SF18   116441 4.935141 -2.790174580  9.790906
ENSG00000184371.13    CSF1     1435 3.485613  1.072198929  9.694764
ENSG00000213626.12     LBH    81606 5.479161  0.001291519  9.594310
ENSG00000173175.14   ADCY5      111 4.154702  0.864838207  8.736459
ENSG00000148671.13   ADIRF    10974 2.534308  2.789357717  8.607413
                        P.Value    adj.P.Val
ENSG00000109610.5  2.297798e-10 1.561357e-06
ENSG00000143839.14 2.330384e-10 1.561357e-06
ENSG00000196352.15 7.293085e-10 3.257578e-06
ENSG00000265107.2  2.029465e-08 6.798707e-05
ENSG00000173546.7  1.566406e-07 4.197968e-04
ENSG00000163762.6  2.019803e-07 4.243956e-04
ENSG00000184371.13 2.229834e-07 4.243956e-04
ENSG00000213626.12 2.533705e-07 4.243956e-04
ENSG00000173175.14 6.844153e-07 1.019018e-03
ENSG00000148671.13 7.979604e-07 1.055968e-03</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">decideTests</span>(twfit))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       treated untreated
Down         6         0
NotSig   13354     13400
Up          40         0</code></pre>
</div>
</div>
</section>
<section id="compare-heatmaps-of-3-pipelines" class="level2">
<h2 class="anchored" data-anchor-id="compare-heatmaps-of-3-pipelines">Compare heatmaps of 3 pipelines</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gplots)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>col <span class="ot">&lt;-</span> <span class="fu">colorpanel</span>(<span class="dv">1000</span>,<span class="st">"blue"</span>,<span class="st">"white"</span>,<span class="st">"red"</span>)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>topgenes1 <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">topTags</span>(qlf, <span class="at">n =</span> <span class="dv">500</span>))</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>topgenes2 <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">topTable</span>(efit, <span class="at">n =</span> <span class="dv">500</span>))</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>topgenes3 <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">topTable</span>(vwfit, <span class="at">n =</span> <span class="dv">500</span>))</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>topgenes_list <span class="ot">&lt;-</span> <span class="fu">list</span>(topgenes1, topgenes2, topgenes3)</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>){</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>  ind <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">rownames</span>(dge<span class="sc">$</span>genes) <span class="sc">%in%</span> topgenes_list[[i]])</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heatmap.2</span>(log2cpm[ind,], </span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>            <span class="at">labRow =</span> dge<span class="sc">$</span>genes<span class="sc">$</span>symbol[ind],</span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a>            <span class="at">labCol =</span> samples<span class="sc">$</span>group,</span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a>            <span class="at">col =</span> col,</span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a>            <span class="at">trace =</span> <span class="st">"none"</span>,</span>
<span id="cb52-15"><a href="#cb52-15" aria-hidden="true" tabindex="-1"></a>            <span class="at">scale =</span> <span class="st">"row"</span>,</span>
<span id="cb52-16"><a href="#cb52-16" aria-hidden="true" tabindex="-1"></a>            <span class="at">dendrogram =</span> <span class="st">"column"</span>,</span>
<span id="cb52-17"><a href="#cb52-17" aria-hidden="true" tabindex="-1"></a>            <span class="at">density.info =</span> <span class="st">"none"</span>,</span>
<span id="cb52-18"><a href="#cb52-18" aria-hidden="true" tabindex="-1"></a>            <span class="at">margins =</span> <span class="fu">c</span>(<span class="dv">7</span>, <span class="dv">5</span>),</span>
<span id="cb52-19"><a href="#cb52-19" aria-hidden="true" tabindex="-1"></a>            <span class="at">lhei =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">10</span>))</span>
<span id="cb52-20"><a href="#cb52-20" aria-hidden="true" tabindex="-1"></a>} <span class="co"># some difference in gene clustering</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-16-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-16-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="gene-set-testing" class="level1">
<h1>Gene set testing</h1>
<section id="edger-pipeline-1" class="level2">
<h2 class="anchored" data-anchor-id="edger-pipeline-1">edgeR pipeline</h2>
<p>Test GO_BP gene set</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>go <span class="ot">&lt;-</span> <span class="fu">goana</span>(qlf, <span class="at">geneid =</span> qlf<span class="sc">$</span>genes<span class="sc">$</span>entrezid) <span class="co"># it needs entrezID</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>gobp <span class="ot">&lt;-</span> <span class="fu">topGO</span>(go, <span class="at">ontology =</span> <span class="st">"BP"</span>)</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a><span class="fu">topGO</span>(go, <span class="at">ontology =</span> <span class="st">"BP"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                                     Term Ont    N  DE
GO:0032501               multicellular organismal process  BP 4343 267
GO:0009888                             tissue development  BP 1262 112
GO:0032502                          developmental process  BP 4011 246
GO:0048856               anatomical structure development  BP 3665 229
GO:0050896                           response to stimulus  BP 5417 299
GO:0007275             multicellular organism development  BP 2943 191
GO:0050793            regulation of developmental process  BP 1579 123
GO:0051239 regulation of multicellular organismal process  BP 1826 135
GO:0030154                           cell differentiation  BP 2641 174
GO:0048869                 cellular developmental process  BP 2642 174
GO:0048513                       animal organ development  BP 1853 135
GO:0009653             anatomical structure morphogenesis  BP 1679 125
GO:0042221                           response to chemical  BP 2348 158
GO:0023052                                      signaling  BP 3778 224
GO:0007154                             cell communication  BP 3835 226
GO:0048731                             system development  BP 2466 163
GO:0016477                                 cell migration  BP  953  82
GO:0042127    regulation of cell population proliferation  BP 1060  88
GO:0007155                                  cell adhesion  BP  854  76
GO:0007166        cell surface receptor signaling pathway  BP 1702 122
                   P.DE
GO:0032501 1.459608e-18
GO:0009888 7.128808e-17
GO:0032502 2.195832e-16
GO:0048856 7.353696e-16
GO:0050896 2.137634e-14
GO:0007275 2.767191e-14
GO:0050793 3.364977e-14
GO:0051239 7.514125e-14
GO:0030154 2.070685e-13
GO:0048869 2.144440e-13
GO:0048513 2.336838e-13
GO:0009653 5.377914e-13
GO:0042221 8.413977e-13
GO:0023052 8.653371e-13
GO:0007154 1.213426e-12
GO:0048731 1.405464e-12
GO:0016477 1.107936e-11
GO:0042127 1.196644e-11
GO:0007155 1.400137e-11
GO:0007166 1.517370e-11</code></pre>
</div>
</div>
</section>
<section id="limma-pipeline" class="level2">
<h2 class="anchored" data-anchor-id="limma-pipeline">limma pipeline</h2>
<p>Test 2 GO_BP gene sets for CaSR.R568</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GO.db)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="fu">select</span>(GO.db,</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">keys =</span> <span class="fu">c</span>(<span class="st">"cell migration"</span>, <span class="st">"cell differentiation"</span>),</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">keytype =</span> <span class="st">"TERM"</span>,</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">column =</span> <span class="st">"GOID"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  TERM       GOID
1       cell migration GO:0016477
2 cell differentiation GO:0030154</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Rkeys</span>(org.Hs.egGO2ALLEGS) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GO:0016477"</span>, <span class="st">"GO:0030154"</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">ids2indices</span>(<span class="fu">as.list</span>(org.Hs.egGO2ALLEGS), <span class="at">id =</span> v<span class="sc">$</span>genes<span class="sc">$</span>entrezid)</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mroast</span>(v, idx, design, <span class="at">contrast =</span> con[,<span class="dv">1</span>]) <span class="co"># both up</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           NGenes  PropDown    PropUp Direction PValue    FDR PValue.Mixed
GO:0016477    953 0.1091291 0.2885624        Up 0.0005 0.0005        0.016
GO:0030154   2642 0.1377744 0.2187737        Up 0.0290 0.0290        0.036
           FDR.Mixed
GO:0016477    0.0315
GO:0030154    0.0360</code></pre>
</div>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">camera</span>(v, idx, design, <span class="at">contrast =</span> con[,<span class="dv">1</span>]) <span class="co"># only cell migration significant</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           NGenes Direction       PValue          FDR
GO:0016477    953        Up 0.0002131288 0.0004262576
GO:0030154   2642        Up 0.0590823820 0.0590823820</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fry</span>(v, idx, design, <span class="at">contrast =</span> con[,<span class="dv">1</span>]) <span class="co"># both up</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           NGenes Direction      PValue          FDR PValue.Mixed    FDR.Mixed
GO:0016477    953        Up 0.000412588 0.0008251761 9.181106e-07 1.836221e-06
GO:0030154   2642        Up 0.019719621 0.0197196214 8.939867e-06 8.939867e-06</code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">barcodeplot</span>(efit<span class="sc">$</span>t[,<span class="dv">1</span>], <span class="at">index =</span> idx<span class="sc">$</span><span class="st">`</span><span class="at">GO:0016477</span><span class="st">`</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="CaSR_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>